processor/src/main/java/org/mapstruct/ap/internal/conversion/DecimalFormatterHelper.java

```diff
@@ -0,0 +1,100 @@
+package org.mapstruct.ap.internal.conversion;
+
+import org.mapstruct.ap.internal.model.common.ConversionContext;
+import org.mapstruct.ap.internal.model.common.Type;
+import org.mapstruct.ap.internal.util.DecimalFormatConstants;
+import org.mapstruct.ap.internal.util.NativeTypes;
+import org.mapstruct.ap.internal.util.Strings;
+
+import java.text.DecimalFormat;
+import java.text.ParseException;
+import java.util.*;
+import java.util.Collections;
+
+/**
+ * @author Ciaran Liedeman
+ */
+public class DecimalFormatterHelper {
+
+    private final Class<?> sourceType;
+    private final Class<?> primitiveType;
+    private final Class<?> wrapperType;
+
+    public DecimalFormatterHelper(final Class<?> sourceType) {
+        this.sourceType = sourceType;
+
+        if ( sourceType.isPrimitive() ) {
+            wrapperType = NativeTypes.getWrapperType( sourceType );
+            primitiveType = sourceType;
+        } else {
+            wrapperType = sourceType;
+            primitiveType = NativeTypes.getPrimitiveType( sourceType );
+        }
+    }
+
+
+    public Set<Type> getToConversionImportTypes(ConversionContext conversionContext) {
+        return java.util.Collections.singleton( conversionContext.getTypeFactory().getType( DecimalFormat.class ) );
+    }
+
+    public Set<Type> getFromConversionImportTypes(ConversionContext conversionContext) {
+        return java.util.Collections.singleton( conversionContext.getTypeFactory().getType( DecimalFormat.class ) );
+    }
+
+    public List<Type> getFromConversionExceptionTypes(ConversionContext conversionContext) {
+        return Collections.singletonList( conversionContext.getTypeFactory().getType( ParseException.class ));
+    }
+
+    public String getFormatterToConversionExpression(ConversionContext conversionContext) {
+        checkIfWeCanFormat();
+
+        StringBuilder sb = new StringBuilder();
+
+        appendDecimalFormatter( sb, conversionContext );
+        sb.append( ".format" );
+        sb.append( "( <SOURCE> )" );
+
+        return sb.toString();
+    }
+
+    public String getFormatterFromConversionExpression(ConversionContext conversionContext) {
+        checkIfWeCanFormat();
+
+        StringBuilder sb = new StringBuilder();
+
+        appendDecimalFormatter( sb, conversionContext );
+        sb.append( ".parse" );
+        sb.append( "( <SOURCE> ).toString()" );
+
+        return getParseTypeExpression( sb.toString() );
+    }
+
+    private void checkIfWeCanFormat() {
+        if ( ! DecimalFormatConstants.NUMBER_TYPES.contains( sourceType )) {
+            throw new IllegalStateException( String.format(
```

Comment:
Is it possible to bring the message up to this level? i'm not in favor of using an `IllegalStateException` for a user error. See for example
```java
ctx.getMessager().printMessage(
                            method.getExecutable(),
                            beanMappingPrism.mirror,
                            Message.BEANMAPPING_NOT_ASSIGNABLE, resultType, method.getResultType()
                        );
```